File: scripts/migration-tests/test-examples.mjs
/** Execute the exact documented TypeScript examples against fault-injecting clients.
 * No credentials, dependencies, network, or live provider sessions are used.
 * Run with Node >=24: node --test scripts/migration-tests/test-examples.mjs
 */
import assert from 'node:assert/strict';
import { mkdtemp, mkdir, readFile, writeFile, rm } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { after, test } from 'node:test';

const doc = new URL('../../plugins/notte-migrate/skills/migrate-to-notte/references/notte-sessions.md', import.meta.url);
const source = await readFile(doc, 'utf8');
const examples = [...source.matchAll(/```typescript\n([\s\S]*?)```/g)].map(m => m[1]);
assert.equal(examples.length, 2, 'Expected Playwright and Puppeteer examples');
const temp = await mkdtemp(path.join(os.tmpdir(), 'notte-migration-examples-'));
after(() => rm(temp, { recursive: true, force: true }));
for (const [name, content] of Object.entries({
  'notte-sdk': 'export class NotteClient { Session(options) { globalThis.fixture.options = options; return globalThis.fixture.session; } }',
  'playwright-core': 'export const chromium = { connectOverCDP: (...args) => globalThis.fixture.connect(...args) };',
  'puppeteer-core': 'export default { connect: (...args) => globalThis.fixture.connect(...args) };',
})) {
  const dir = path.join(temp, 'node_modules', name);
  await mkdir(dir, { recursive: true });
  await writeFile(path.join(dir, 'package.json'), JSON.stringify({ name, type: 'module', exports: './index.js' }));
  await writeFile(path.join(dir, 'index.js'), content);
}
const modules = [];
for (let i = 0; i < examples.length; i++) {
  const file = path.join(temp, `example-${i}.mts`);
  await writeFile(file, examples[i]);
  modules.push(await import(pathToFileURL(file)));
}

function fixture(failures, { emptyPage = false, missingContext = false } = {}) {
  const calls = [];
  const errors = Object.fromEntries(failures.map(stage => [stage, new Error(`${stage} failed`)]));
  const hit = stage => { calls.push(stage); if (errors[stage]) throw errors[stage]; };
  const page = { goto: async () => hit('goto'), title: async () => { hit('title'); return 'Expected application title'; } };
  const context = { pages: () => emptyPage ? [] : [page], newPage: async () => { hit('newPage'); return page; } };
  const browser = {
    contexts: () => missingContext ? [] : [context],
    defaultBrowserContext: () => missingContext ? undefined : context,
    close: async () => hit('clientCleanup'),
    disconnect: async () => hit('clientCleanup'),
  };
  const session = {
    start: async () => hit('start'),
    cdpUrl: async () => { hit('cdp'); return 'wss://fixture.invalid/browser'; },
    stop: async () => hit('stop'),
    // This double follows the SDK's documented callback/stop error precedence.
    async use(callback) {
      await this.start();
      let failed = false;
      try { return await callback(this); }
      catch (e) { failed = true; console.warn(`Session exiting because of exception: ${String(e)}`); throw e; }
      finally { try { await this.stop(); } catch (e) { if (!failed) throw e; } }
    },
  };
  return { calls, errors, session, connect: async () => { hit('connect'); return browser; } };
}

for (const [i, driver] of ['Playwright', 'Puppeteer'].entries()) {
  for (const failures of [[], ['start'], ['cdp'], ['connect'], ['goto'], ['title'], ['clientCleanup'], ['stop'], ['goto', 'clientCleanup'], ['goto', 'stop'], ['goto', 'clientCleanup', 'stop'], ['connect', 'stop']]) {
    test(`${driver}: ${failures.join(' + ') || 'success'}`, async () => {
      const f = fixture(failures);
      globalThis.fixture = f;
      if (failures.length) {
        await assert.rejects(modules[i].readTitle('https://fixture.invalid'), error => error === f.errors[failures[0]], 'Preserve the original failure');
      } else {
        assert.equal(await modules[i].readTitle('https://fixture.invalid'), 'Expected application title');
      }
      assert.equal(f.calls.filter(c => c === 'stop').length, failures.includes('start') ? 0 : 1, 'Stop every successfully started session');
      const connected = !failures.some(c => ['start', 'cdp', 'connect'].includes(c));
      assert.equal(f.calls.filter(c => c === 'clientCleanup').length, connected ? 1 : 0);
      assert.equal(f.options.max_duration_minutes, 5);
      assert.equal(f.options.idle_timeout_minutes, 2);
    });
  }
  test(`${driver}: reuse default context when initial page is absent`, async () => {
    const f = fixture([], { emptyPage: true }); globalThis.fixture = f;
    assert.equal(await modules[i].readTitle('https://fixture.invalid'), 'Expected application title');
    assert.equal(f.calls.filter(c => c === 'newPage').length, 1);
    assert.equal(f.calls.filter(c => c === 'stop').length, 1);
  });
  test(`${driver}: missing context still cleans up`, async () => {
    const f = fixture([], { missingContext: true }); globalThis.fixture = f;
    await assert.rejects(modules[i].readTitle('https://fixture.invalid'));
    assert.ok(f.calls.includes('clientCleanup'));
    assert.ok(f.calls.includes('stop'));
  });
}

for (const [i, driver] of ['Playwright', 'Puppeteer'].entries()) {
  test(`${driver}: transport errors are not logged by the example`, async () => {
    const f = fixture(['connect']);
    f.errors.connect.message = 'connect failed: wss://fixture.invalid?apiKey=synthetic-secret';
    globalThis.fixture = f;
    const original = { warn: console.warn, error: console.error, log: console.log };
    const output = [];
    for (const key of Object.keys(original)) console[key] = (...args) => output.push(args.join(' '));
    try {
      await assert.rejects(modules[i].readTitle('https://fixture.invalid'), error => error === f.errors.connect);
      assert.equal(output.length, 0, 'Caller owns sanitized error reporting');
    } finally {
      Object.assign(console, original);
    }
  });
}

File: scripts/migration-tests/test-examples.py
"""Execute the documented Python example with fault-injecting clients, offline."""

from __future__ import annotations

import re
import sys
import types
import unittest
from pathlib import Path
from unittest.mock import patch

DOC = (
    Path(__file__).resolve().parents[2]
    / "plugins/notte-migrate/skills/migrate-to-notte/references/notte-sessions.md"
)
EXAMPLES = re.findall(r"```python\n(.*?)```", DOC.read_text(), re.DOTALL)
if len(EXAMPLES) != 1:
    raise RuntimeError("Expected one Python lifecycle example")


class ExampleTests(unittest.TestCase):
    def run_case(
        self,
        failures=(),
        *,
        empty_page=False,
        missing_context=False,
        cancellation=False,
    ):
        calls = []
        errors = {name: RuntimeError(f"{name} failed") for name in failures}
        if cancellation:
            errors["goto"] = KeyboardInterrupt()

        def hit(name):
            calls.append(name)
            if name in errors:
                raise errors[name]

        page = types.SimpleNamespace(
            goto=lambda url: hit("goto"),
            title=lambda: (hit("title"), "Expected application title")[1],
        )
        context = types.SimpleNamespace(
            pages=[] if empty_page else [page],
            new_page=lambda: (hit("new_page"), page)[1],
        )
        browser = types.SimpleNamespace(
            contexts=[] if missing_context else [context],
            close=lambda: hit("client_cleanup"),
        )
        session = types.SimpleNamespace(
            start=lambda: hit("start"),
            stop=lambda: hit("stop"),
            cdp_url=lambda: (hit("cdp"), "wss://fixture.invalid/browser")[1],
        )
        playwright = types.SimpleNamespace(
            chromium=types.SimpleNamespace(
                connect_over_cdp=lambda url: (hit("connect"), browser)[1]
            ),
            stop=lambda: hit("playwright_stop"),
        )
        factory = types.SimpleNamespace(
            start=lambda: (hit("playwright_start"), playwright)[1]
        )
        notte_module = types.ModuleType("notte_sdk")
        notte_module.NotteClient = lambda: types.SimpleNamespace(
            Session=lambda **options: session
        )
        playwright_module = types.ModuleType("playwright.sync_api")
        playwright_module.sync_playwright = lambda: factory
        namespace = {}
        with patch.dict(
            sys.modules,
            {
                "notte_sdk": notte_module,
                "playwright": types.ModuleType("playwright"),
                "playwright.sync_api": playwright_module,
            },
        ):
            exec(compile(EXAMPLES[0], str(DOC), "exec"), namespace)  # noqa: S102 - execute repository-owned examples
            if failures:
                expected = errors[failures[0]]
                with self.assertRaises(type(expected)) as caught:
                    namespace["read_title"]("https://fixture.invalid")
                self.assertIs(
                    caught.exception,
                    expected,
                    "Cleanup must not replace the primary error",
                )
            elif missing_context:
                with self.assertRaises((IndexError, RuntimeError)):
                    namespace["read_title"]("https://fixture.invalid")
            else:
                self.assertEqual(
                    namespace["read_title"]("https://fixture.invalid"),
                    "Expected application title",
                )
        self.assertEqual(calls.count("stop"), 0 if "start" in failures else 1)
        connected = not any(
            x in failures for x in ("start", "playwright_start", "cdp", "connect")
        )
        self.assertEqual(calls.count("client_cleanup"), int(connected))
        pw_started = not any(x in failures for x in ("start", "playwright_start"))
        self.assertEqual(calls.count("playwright_stop"), int(pw_started))
        if empty_page:
            self.assertEqual(calls.count("new_page"), 1)

    def test_empty_page_reuses_default_context(self):
        self.run_case(empty_page=True)

    def test_missing_context_cleans_up(self):
        self.run_case(missing_context=True)

    def test_keyboard_interrupt_preserved_through_cleanup_failure(self):
        self.run_case(
            ("goto", "client_cleanup", "stop", "playwright_stop"), cancellation=True
        )


for failures in [
    (),
    ("start",),
    ("playwright_start",),
    ("cdp",),
    ("connect",),
    ("goto",),
    ("title",),
    ("client_cleanup",),
    ("stop",),
    ("playwright_stop",),
    ("goto", "client_cleanup"),
    ("goto", "stop"),
    ("goto", "client_cleanup", "stop", "playwright_stop"),
    ("connect", "stop"),
    ("client_cleanup", "stop"),
    ("stop", "playwright_stop"),
]:

    def case(self, failures=failures):
        self.run_case(failures)

    setattr(ExampleTests, "test_" + ("_and_".join(failures) or "success"), case)

if __name__ == "__main__":
    unittest.main(verbosity=2)
